components/formatters/src/duration/duration-formatter.ts

```diff
@@ -0,0 +1,249 @@
+  [DtTimeUnit.MINUTE, 60 * 1000],
+  [DtTimeUnit.SECOND, 1000],
+  [DtTimeUnit.MILLISECOND, 1],
+  [DtTimeUnit.MICROSECOND, 0.001],
+  [DtTimeUnit.NANOSECOND, 0.000001],
+]);
+
+const CONVERSIONUNITS = 3;
+
+/**
+ * Formats a numeric value to a duration
+ * @param duration numeric time value
+ * @param formatMethod the formatting method
+ * @param outputUnit dtTimeUnit | undefined value describing the unit to which it should format e.g to seconds
+ * @param inputUnit dtTimeUnit value describing which unit the duration is in
+ */
+export function formatDuration(
+  duration: number,
+  formatMethod: string | undefined,
+  outputUnit: DtTimeUnit | undefined,
+  inputUnit: DtTimeUnit = DtTimeUnit.MILLISECOND,
+): DtFormattedValue | string {
+  const inputData: SourceData = {
+    input: duration,
+    unit: inputUnit,
+  };
+  let formattedData: FormattedData;
+  let result: Map<DtTimeUnit, string> | undefined;
+
+  if (duration <= 0 && formatMethod !== DurationMode.PRECISE) {
+    // Smaller than 0 duration impossible
+    return new DtFormattedValue(inputData, {
+      transformedValue: inputData.input,
+      displayValue: '< 1',
+      displayUnit: inputUnit,
+      displayWhiteSpace: false,
+    });
+  } else {
+    if (!formatMethod) {
+      formatMethod = DurationMode.DEFAULT;
+    }
+    if (formatMethod === DurationMode.PRECISE) {
+      result = transformResultWithOutputUnit(
+        duration,
+        inputUnit,
+        outputUnit,
+        formatMethod,
+      );
+    } else if (formatMethod === DurationMode.DEFAULT) {
+      if (outputUnit) {
+        result = transformResultWithOutputUnit(
+          duration,
+          inputUnit,
+          outputUnit,
+          formatMethod,
+        );
+      } else {
+        result = transformResult(duration, inputUnit, formatMethod);
+      }
+    } else {
+      result = transformResult(duration, inputUnit, formatMethod);
+    }
+
+    // Return NO_DATA when inputUnit is invalid
+    if (CONVERSION_FACTORS_TO_MS.get(inputUnit) === undefined) {
+      return NO_DATA;
+    }
+    if (result === undefined) {
+      return NO_DATA;
+    }
+    let resultString = '';
+    result.forEach((value, key) => {
+      resultString = `${resultString}${value} ${key} `;
+    });
+    resultString = resultString.trim();
+    formattedData = {
+      transformedValue: inputData.input,
+      displayValue: resultString,
+      displayUnit: undefined,
+      displayWhiteSpace: false,
+    };
+    return new DtFormattedValue(inputData, formattedData);
+  }
+}
+
+/**
+ * Calculates output duration in either "DEFAULT" or "CUSTOM" mode.
+ * If precision is DEFAULT or undefined then display a maximum of three units, but
+ * when precision is a number(custom), then display that amount of units.
+ * @param duration numeric time value
+ * @param inputUnit dtTimeUnit value describing which unit the duration is in
+ * @param formatMethod the formatting method
+ */
+function transformResult(
```

Comment:
I think we should export the function here and create a unit test for this function alone. We have a lot of magic happening in all these functions and I want to test them in isolation